components/brave_webtorrent/browser/magnet_protocol_handler.cc

```diff
@@ -0,0 +1,150 @@
+  GURL extension_page_url(base::StrCat(
+      {extensions::kExtensionScheme, "://", brave_webtorrent_extension_id,
+       "/extension/brave_webtorrent.html?%s"}));
+  std::string translatedSpec(extension_page_url.spec());
+  base::ReplaceFirstSubstringAfterOffset(
+      &translatedSpec, 0, "%s", base::EscapeQueryParamValue(url.spec(), true));
+  return GURL(translatedSpec);
+}
+
+GURL TranslateTorrentUIURLReversed(const GURL& url) {
+  GURL translatedURL(base::UnescapeURLComponent(
+      url.query(),
+      base::UnescapeRule::URL_SPECIAL_CHARS_EXCEPT_PATH_SEPARATORS |
+          base::UnescapeRule::PATH_SEPARATORS));
+  GURL::Replacements replacements;
+  replacements.SetRefStr(url.ref_piece());
+  return translatedURL.ReplaceComponents(replacements);
+}
+
+bool HandleTorrentURLReverseRewrite(GURL* url,
+                                    content::BrowserContext* browser_context) {
+  if (url->SchemeIs(extensions::kExtensionScheme) &&
+      url->host() == brave_webtorrent_extension_id &&
+      url->ExtractFileName() == "brave_webtorrent.html" &&
+      GURL(url->query()).SchemeIsHTTPOrHTTPS()) {
+    *url = TranslateTorrentUIURLReversed(*url);
+    return true;
+  }
+
+  return false;
+}
+
+bool HandleTorrentURLRewrite(GURL* url,
+                             content::BrowserContext* browser_context) {
+  if (!IsWebtorrentEnabled(browser_context)) {
+    return false;
+  }
+
+  // The HTTP/HTTPS URL could be modified later by the network delegate if the
+  // mime type matches or .torrent is in the path.
+  // Handle http and https here for making reverse_on_redirect to be true in
+  // BrowserURLHandlerImpl::RewriteURLIfNecessary to trigger ReverseURLRewrite
+  // for updating the virtual URL.
+  if (url->SchemeIsHTTPOrHTTPS() ||
+      (url->SchemeIs(extensions::kExtensionScheme) &&
+       url->host() == brave_webtorrent_extension_id &&
+       url->ExtractFileName() == "brave_webtorrent.html" &&
+       GURL(url->query()).SchemeIsHTTPOrHTTPS())) {
+    return true;
+  }
+
+  return false;
+}
+
+bool HandleMagnetURLRewrite(GURL* url,
+                            content::BrowserContext* browser_context) {
+  if (IsWebtorrentEnabled(browser_context) && IsMagnetProtocol(*url)) {
+    *url = TranslateMagnetURL(*url);
+    return true;
+  }
+
+  return false;
+}
+
+bool HandleMagnetProtocol(const GURL& url,
+                          content::WebContents::Getter web_contents_getter,
+                          ui::PageTransition page_transition,
+                          bool has_user_gesture,
+                          bool is_in_fenced_frame_tree,
+                          const absl::optional<url::Origin>& initiating_origin,
+                          content::WeakDocumentPtr initiator_document) {
+  if (!IsMagnetProtocol(url)) {
+    return false;
+  }
+
+  // Handle subframe magnet links only if a user gesture is present.
+  if (!ui::PageTransitionIsMainFrame(page_transition) && !has_user_gesture) {
```

Comment:
cc @fmarier @fallaciousreasoning 

this is a new thing here. Without it we just blindly open Webtorrent if an iframe src points to `magnet:` url. This looked really crazy, so I added this condition. We can fully disallow magnet redirects from subframes, but this is for another task which may also include other Webtorrent-related changes.

Reply:
That's a bit scary :fearful: